### PR TITLE
Refactoring `getNavigationTree` to include folders

### PR DESCRIPTION
**Pull request by Saintpreston**
[**Asana Task**](https://app.asana.com/0/home/1204780532876376/1204807557661947) 🔗

### What

- Restructures exported member, `categories` from `swingset/meta` 
- Updates rendering logic in `hashicorp-theme` to match new `NavigationTree` data structure.

### Why

- Added functionality for `Folder` this way themes can easier integrate multifaceted components or sub components. 
- Allows for easier long term goal of sorting and searching functionality.

### How

- Rewrote `getNavigationTree`
- Added new `Node` types within tree

### CODE REVIEW

File: packages/swingset-theme-hashicorp/src/components/side-nav/category.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      </section>
+    </li>
+  )
+}
+//Swap this out for already existing heading
+function CategoryHeading({ children }: { children: string }) {
+  return <h4>{children.toUpperCase()}</h4>
+}
+
+function ComponentList({
```
Comment (Saintpreston) [inline packages/swingset-theme-hashicorp/src/components/side-nav/category.tsx:26]: This should be renamed to `CategoryList` or `CategoryItems` for consistency with the other components

File: packages/swingset-theme-hashicorp/src/components/side-nav/index.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+function SideNavBar(props: SideNavBarProps) {
+  const { categories } = props
+
+  
+ 
+
+  const categoriesArr = Object.entries(categories)
+  console.log(categoriesArr)
```
Comment (Saintpreston) [inline packages/swingset-theme-hashicorp/src/components/side-nav/index.tsx:48]: remove log

File: packages/swingset/src/get-nav-tree.ts [hunk 3]
```diff
@@ -1,49 +1,95 @@
...
+    const categoryTitle = entity.navigationData?.category || 'default'
+    const folderTitle = entity.navigationData?.folder
+    const hasFolder = !!folderTitle
+
+    const storedCategory =
+      categories.has(categoryTitle) && categories.get(categoryTitle)
+
+     
+
+    if (!!storedCategory) {
```
Comment (BRKalow) [inline packages/swingset/src/get-nav-tree.ts:55]: I would potentially split out the category creation (if not exists) from the folder handling. I think the flow would look like (in plain english):

```
check if category is stored

if category is not stored, create the category object

then, handle folder logic
```

This dedupes some of the folder logic between these two branches.

(optional) I also wonder if it'd be easier to build up the category map and folder map separately, and then after we're done iterating over all of the entities we can make sure each folder gets added to its proper category. that way we don't need to check each of a category's children for the correct folder when we're trying to add an element to it.

File: packages/swingset-theme-hashicorp/src/components/side-nav/category.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        const hasChildren = (item.children?.length as number) > 0
+
+        if (hasChildren) {
+          return (
+            <li key={item.title}>
+              <LinkItem to={item.slug} title={item.title} />
+              <ComponentList
+                isNested
+                items={item.children as unknown[] as ComponentNode[]}
```
[the remark this answers is not in the document]
Comment (BRKalow) [reply packages/swingset-theme-hashicorp/src/components/side-nav/category.tsx:61]: Want to add a `// TODO` here linking to the issue so we know to clean this up as part of addressing that issue?

File: packages/swingset/src/types.ts [hunk 5]
```diff
@@ -40,3 +40,25 @@ export type EvaluatedEntity<T extends Entity = ComponentEntity | DocsEntity> =
...
+  __type: 'folder'
+  title: string
+  parentCategory: string
+  children: ComponentNode[]
+}
+
+export type NavigationNode = ComponentNode | FolderNode
+
+export type CategoryNode = {
+  type: 'category'
```
Comment (BRKalow) [inline packages/swingset/src/types.ts:59]: Is `__type` the name we want to use? (I think yes based on your other comments)

```suggestion
  __type: 'category'
```

File: packages/swingset/src/get-nav-tree.ts [hunk 6]
```diff
@@ -1,49 +1,86 @@
...
+    }
 
-    result[category].sort((a, b) => (a.slug > b.slug ? 1 : -1))
+    //if node doesnt belong in folder, add node
+    storedCategory.children.push(componentNode)
   }
 
-  return result
+  const tree = Object.fromEntries(categories)
+  return tree
```
Comment (BRKalow) [inline packages/swingset/src/get-nav-tree.ts:85]: We want to return `NavigationTree` here, right? That's an array of `CategoryNode`, so we don't want an object here.

This might work (builds an array from the `Map`'s values):
```ts
Array.from(categories.values())
```

File: packages/swingset-theme-hashicorp/src/components/side-nav/index.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+function SideNavigation(props: SideNavBarProps) {
+  const { categories } = props
+
+  
+
+  const categoriesJSX = Object.values(categories).map(
+    (category) => (
+      <Category title={category.title} key={category.title} items={category.children} />
+    )
+  )
```
Comment (BRKalow) [inline packages/swingset-theme-hashicorp/src/components/side-nav/index.tsx:17]: I _think_ `Object.values()` will be unnecessary here if we fix the return value of `getNavigationTree`. In that case, we can use `categories.map()`.

File: packages/swingset-theme-hashicorp/src/components/side-nav/index.tsx [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { NavigationTree } from 'swingset/types'
+import Category from './category'
+
+type SideNavBarProps = {
+  categories: NavigationTree
+}
+
+function SideNavigation(props: SideNavBarProps) {
+  const { categories } = props
```
Comment (BRKalow) [inline packages/swingset-theme-hashicorp/src/components/side-nav/index.tsx:9]: nit, we can destructure this:
```suggestion
function SideNavigation({ categories }: SideNavBarProps) {
```

File: packages/swingset-theme-hashicorp/src/components/side-nav/link-item.tsx [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import { Link } from "../link"
+import { cx } from "class-variance-authority"
+
+function LinkItem({ title, to }: Record<'title' | 'to', string>) {
```
Comment (BRKalow) [inline packages/swingset-theme-hashicorp/src/components/side-nav/link-item.tsx:4]: The use of `Record` is clever here, but I think it'd be clearer if we used a more conventional type definition:
```suggestion
type LinkItemProps = {
	title: string
	to: string
}

function LinkItem({ title, to }: LinkItemProps) {
```

File: packages/swingset-theme-hashicorp/src/components/side-nav/category.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+}) {
+  return (
+    <ul
+      className={cx(
+        'ss-mt-2 ss-space-y-1',
+        isNested && 'ss-ml-2 ss-border-l ss-border-faint ss-mt-0'
+      )}
+    >
+      {items.map((item) => {
+        const isFolder = item.__type === 'folder'
```
Comment (BRKalow) [inline packages/swingset-theme-hashicorp/src/components/side-nav/category.tsx:45]: This could be a good candidate for using a switch statement on `__type`. It's stylistic really, but I think it would make the logic of this block easier to understand.

File: packages/swingset/src/get-nav-tree.ts [hunk 11]
```diff
@@ -1,49 +1,86 @@
...
+          childEntity.componentPath === componentEntity.componentPath
+      )
 
-    return entity
-  })
+      return componentEntity
+    }
+  )
+
+  //TODO: Account for duplicate folder names [category]/[folder]
```
Comment (BRKalow) [inline packages/swingset/src/get-nav-tree.ts:27]: Just noting that this is applicable for folders, not necessarily categories

File: packages/swingset/src/get-nav-tree.ts [hunk 12]
```diff
@@ -1,49 +1,86 @@
-import { ComponentEntity, DocsEntity } from './types.js'
+import {
+  ComponentEntity,
+  DocsEntity,
+  ComponentNode,
+  CategoryNode,
+} from './types.js'
 
 export function getNavigationTree(entities: (ComponentEntity | DocsEntity)[]) {
```
Comment (BRKalow) [inline packages/swingset/src/get-nav-tree.ts:8]: Some unit tests for this function could be helpful! It'd make it easier for us to iterate on the logic going forward too. We can do this as a follow-up.

**Review by BRKalow — COMMENTED** [review]
Looking good! I think the component breakdown you've come up with is sensible. There seems to be a mismatch between what we're returning from `getNavigationTree` and the type we're expecting.

File: packages/swingset/src/types.ts (hunk 5 above)
Comment (Saintpreston) [reply packages/swingset/src/types.ts:59]: Originally the `Omit<...> & {type: '....'}` changed it but I saw in the first review you requested to remove the Omit.

I'm impartial on the naming but think `type` felt more consistent with the other properties.

Comment (BRKalow) [reply packages/swingset/src/types.ts:59]: The other two nodes use `__type`, so I'd say let's pick one or the other and be consistent between each interface 👍

Comment (Saintpreston) [reply packages/swingset/src/types.ts:59]: Will stick with what's already there then, since `__type` is used across a few other files already
